doc/style.md

```diff
@@ -0,0 +1,274 @@
+
+```scala
+val queueOut = Queue(
+  Arbitrate.byRoundRobin(
+    Queue(a), // depth assumed to be 1
+    Queue(b, depth=3),
+    Queue(c, depth=4)
+  ),
+  depth=10
+)
+```
+## Bundles
+
+Bundles are used for interfaces and structs.
+
+## Naming Conventions
+
+Chisel follows the [Scala Naming Conventions]
+(http://docs.scala-lang.org/style/naming-conventions.html). In general, Chisel
+code should use CamelCase for naming (ie. the first letter of each word is
+capitalized except sometimes the first word).
+
+### Why CamelCase instead of Snake\_Case?
+
+The compiler inserts underscores when splitting Chisel/FIRRTL aggregate types
+into Verilog types. The compiler uses underscores to preserve the original
+structure of the data in the resulting Verilog. Because of the special meaning
+of underscores in Chisel-generated Verilog, their use in naming is **strongly**
+discouraged.
+
+Consider the following Chisel code:
+
+```scala
+val msg = Wire(new Bundle {
+  val valid = Bool()
+  val addr = UInt(32)
+  val data = UInt(64)
+})
+val msg_rec = Wire(Bool())
+```
+
+Which compiles to the Verilog:
+
+```verilog
+wire  msg_valid;
+wire [31:0] msg_addr;
+wire [63:0] msg_data;
+wire  msg_rec;
+```
+
+The Verilog maintains the structure of the original aggregate wire `msg`.
+However, because we named another variable `msg_rec`, it appears in the Verilog
+as if `msg` had 4 fields instead of its actual 3! If we instead follow the
+lowerCamelCase for values naming convention, the resulting Verilog makes more
+sense:
+
+```scala
+val msg = Wire(new Bundle {
+  val valid = Bool()
+  val addr = UInt(32)
+  val data = UInt(64)
+})
+val msgRec = Wire(Bool())
+```
+
+And its resulting Verilog:
+
+```verilog
+wire  msg_valid;
+wire [31:0] msg_addr;
+wire [63:0] msg_data;
+wire  msgRec;
+```
+
+Much better.
+
+### Constants
+
+Unlike the Google Java style, constants use UpperCamelCase, which is in line
+with the official [Scala Naming
+Conventions](https://docs.scala-lang.org/style/naming-conventions.html).
+Constants are final fields (val or object) whose contents are deeply immutable
+and belong to a package object or an object. Examples:
+
+```scala
+// Constants
+object Constants {
+  val Number = 5
+  val Names = "Ed" :: "Ann" :: Nil
+  val Ages = Map("Ed" -> 35, "Ann" -> 32)
+}
+
+// Not constants
+class NonConstantsInClass {
+  val inClass: String = "in-class"
+}
+
+object nonConstantsInObject {
+  var varString = "var-string"
+  val mutableCollection: scala.collection.mutable.Set[String]
+  val mutableElements = Set(mutable)
+}
+```
+
+### UpperCamelCase vs. lowerCamelCase
+
+There is more than one reasonable way to covert English prose into camel case.
+We follow the convention defined in the [Google Java style
+guide](https://google.github.io/styleguide/javaguide.html#s5.3-camel-case).
+
+Note that the casing of the original words is almost entirely disregarded.
+Example:
+
+Prose form     | UpperCamelCase | lowerCamelCase | Incorrect
```

Comment:
Its a bit confusing why you have both versions of camel case so it might be worth reiterating that different kinds of variables use the two different styles.

Reply:
I put the this section last in the naming section. So people will know of examples of when to use lowerCamelCase and when to use UpperCamelCase.

Reply:
Consider adding a strikethrough on the incorrect entries?
Also, I think the main potentially nonobvious rule here is that acronyms are treated as a word for the purposes of camelcase, consider reproducing that here?